atsign-foundation/at_tools
# feat: Add additional encryption-related fields to Metadata

**- What I did**
feat: Add additional encryption-related fields to Metadata

**- How I did it**
* feat: Add encAlgo to metadata, thus allowing multiple algorithms to be used, not just the legacy default of AES/SIC/PKCS7Padding when encrypting data or RSA when encrypting symmetric keys
* feat: Add encKeyName to metadata, thus enabling use of multiple symmetric encryption keys, not just the legacy default
* feat: Add ivNonce to metadata
* feat: Add skeEncKeyName to metadata, thus enabling shared symmetric keys, when they are passed inline in the metadata, to be encrypted with any available public keys, not just the default. (skeEncKeyName is the key name of the public key that was used to encrypt the `sharedKeyEnc` value)
* feat: Add skeEncAlgo to metadata, to go along with skeEncKeyName just as encAlgo goes with encKeyName.
* feat: Added the new metadata fields as optional params to the `update` and `update:meta` verb syntax
* feat: Added new metadata fields to the various parts of the UpdateVerbBuilder
* docs: Added code documentation for new fields
* docs: Cleaned up some other code documentation
* test: Added unit tests to exercise the new metadata fields for building and parsing commands
* test: Added unit tests to cover already-existing functionality which was not being tested
* fix: Found some bugs as a result of adding those unit tests for already-existing functionality, and fixed them
* fix: Some existing metadata fields (dataSignature, encoding) were not previously included as optional params to the `update:meta` verb syntax. They are now, as well as the new metadata fields.
* fix: the `isJson` block of code in UpdateVerbBuilder's buildCommand function was not setting pubKeyCS in the updateParams; it is now, as well as the new metadata fields
* fix: UpdateVerbBuilder.getBuilder was not parsing pubKeyCS, sharedKeyStatus nor encoding from the supplied command. It is now, as well as the new metadata fields
* refactor: Split the verb syntax for update and update:meta over multiple lines so that they are easier to read
* refactor: Changed the verb syntax for update and update:meta so that they consistently have the delimiting ':' at the start of the regex fragments for parameters rather than at the end.
* refactor: Removed extraneous '?:' from regexes, we do not have any non-capturing groups

**- How to verify it**
- [x] Checks pass for this PR
- [x] Backwards compatibility
    Other packages which depend on at_commons should not need to make any change in order for their checks to pass. So we need checks for at_client_sdk, at_libraries and at_server repos all to pass when they are using this at_commons branch via dependency override
  - [x] at_client_sdk https://github.com/atsign-foundation/at_client_sdk/pull/853
  - [x] at_libraries https://github.com/atsign-foundation/at_libraries/pull/290
  - [x] at_server https://github.com/atsign-foundation/at_server/pull/1174

**- Description for the changelog**
feat: Add additional encryption-related fields to Metadata
